proposals/0043-prefetch.md

```diff
@@ -0,0 +1,147 @@
+- Start Date: 2023-11-01
+- Reference Issues: https://github.com/withastro/roadmap/issues/754
+- Implementation PR: https://github.com/withastro/astro/pull/8951
+
+# Summary
+
+Deprecate `@astrojs/prefetch` and provide first-class prefetch support in Astro.
+
+# Example
+
+```html
+<!-- prefetch defaults to hover -->
+<a href="/foo" data-astro-prefetch>foo</a>
+
+<!-- prefetch on tap (touchstart, mousedown events) -->
+<a href="/foo" data-astro-prefetch="tap">foo</a>
+
+<!-- prefetch on hover (focusin, focusout, mouseenter, mouseleave events) -->
+<a href="/foo" data-astro-prefetch="hover">foo</a>
+
+<!-- prefetch on enter viewport (IntersectionObserver) -->
+<a href="/foo" data-astro-prefetch="viewport">foo</a>
+
+<!-- disable prefetch if prefetch all by default -->
+<a href="/foo" data-astro-prefetch="false">foo</a>
+```
+You can also configure options in `astro.config.js`:
+
+```js
+prefetch: {
+  // Whether to prefetch all links by default. same as adding `data-astro-prefetch` to all links.
+  // Default false, but true if view transitions is used. User can explicitly set this to false
+  // to disable for view transitions. (I tried coming up a better name)
+  prefetchAll: true,
+  // Default prefetch strategy for `data-astro-prefetch` (no value specified).
+  defaultStrategy: 'hover' // accepts 'tap' and 'viewport' too
+}
+```
+
+Different prefetch strategies have different behaviours and opinions:
+- `tap`: Calls `fetch()` on `touchstart` or `mousedown` events (they are called before `click` event)
+- `hover`: Calls `fetch()` on `focusin`+`focusout` or `mouseenter`+`mouseleave` events. Hover detection kicks in after 80ms delay.
+- `viewport`: Creates a `<link rel="prefetch">` on enter viewport. It has lower priority than `fetch()` to not clog up the request. Intersection detection kicks in after 300ms.
+
+Notes: `hover` and `viewport` only works on `<a />` tags on initial page load (and view transition page load) due to limitations of the events. Unless we watch the entire DOM with MutationObserver but it's not performant.
+
+---
+
+For programmatic usage (only if `prefetch` config is enabled):
```

Comment:
Should we have a warning if someone attempts to use the programmatic API without having enabled it in their config?

Reply:
Should make sure this doesn't happen for the ViewTransitions case.